File: app/loops.go
package main

// Loop is the top-level construct that replaces Chain in the loop-engineering
// model. A Loop has structured Intent, a Body that is a flowchart of Nodes
// connected by Edges, a convergence predicate that decides when to stop, and
// stop conditions that bound runaway iteration.
//
// A 1-iteration Loop with a trivially-true convergence predicate is
// behaviorally identical to today's Chain run — backwards compat is automatic
// once the runner learns to consume Loop.
//
// These types are scaffolding (Phase A1): defined but not wired into the
// runner yet. See the plan at
// ~/.claude/plans/okay-the-idea-of-replicated-popcorn.md for the full
// design and phase breakdown.
type Loop struct {
	ID                   string          `json:"id"`
	Name                 string          `json:"name"`
	Description          string          `json:"description"`
	Intent               Intent          `json:"intent"`
	Body                 LoopBody        `json:"body"`
	MaxIterations        int             `json:"max_iterations"`
	ConvergencePredicate string          `json:"convergence_predicate"` // JMESPath, bool
	ConvergenceMetric    string          `json:"convergence_metric"`    // JMESPath, number
	StopConditions       []StopCondition `json:"stop_conditions"`
	Permissions          string          `json:"permissions"` // "inherit" | "default" | "strict"
	TemplateSnapshot     *TemplateSnapshot `json:"template_snapshot,omitempty"`
	Cwd                  string          `json:"cwd"`
	WorkspaceProfile     string          `json:"workspace_profile"`
	CreatedAt            string          `json:"created_at"`
	UpdatedAt            string          `json:"updated_at"`
}

// Intent is the structured "what does done look like" attached to a Loop.
// Convergence is the same expression used by Loop.ConvergencePredicate —
// they're the same data viewed from two angles. A template that fails to
// declare Convergence cannot be loaded.
type Intent struct {
	Outcome      string             `json:"outcome"`
	Verification []string           `json:"verification"`
	NonGoals     []string           `json:"non_goals"`
	Convergence  string             `json:"convergence"` // JMESPath, bool
	Escalation   []EscalationClause `json:"escalation"`
}

// EscalationClause routes an out-of-bounds Loop state to a handler: emit a
// human-attention envelope, skip the iteration, or fail the Loop outright.
type EscalationClause struct {
	Predicate string `json:"predicate"` // JMESPath, bool
	Action    string `json:"action"`    // "human" | "skip" | "fail"
}

// StopCondition is a hard cap evaluated between iterations. If any matches,
// the Loop stops with a corresponding status.
type StopCondition struct {
	Predicate string `json:"predicate"` // JMESPath, bool against envelope
	Reason    string `json:"reason"`    // operator-facing tag, e.g. "diff_too_large"
}

// LoopBody is the flowchart that runs each iteration. Nodes are units of
// execution; Edges connect them and may carry predicates that select branches
// based on the envelope emitted by the source Node.
type LoopBody struct {
	Nodes []Node `json:"nodes"`
	Edges []Edge `json:"edges"`
}

// Node is one unit of execution inside a Loop body. Kind selects the runtime
// shape (agent invocation, brainbox playbook, join of upstream branches, human
// review pause, scheduled wait). Executor selects the runtime backend for
// agent/playbook nodes — host-cli is wired today; brainbox-session and
// a2a-remote are forward-compat slots.
//
// Requires lists permission scopes the Node needs (consulted only in the
// "strict" permission tier; in "default" tier, destructive scopes still
// require explicit listing here).
type Node struct {
	ID         string   `json:"id"`
	Kind       string   `json:"kind"`     // "agent" | "playbook" | "join" | "human" | "schedule"
	Executor   string   `json:"executor"` // "host-cli" | "brainbox-session" | "a2a-remote"
	Role       string   `json:"role"`    // references brainbox/agents/roles/*.md by name
	AgentID    string   `json:"agent_id"`
	PlaybookID string   `json:"playbook_id"`
	Prompt     string   `json:"prompt"`
	Requires   []string `json:"requires"`
	TimeoutMs  int      `json:"timeout_ms"`
}

// Edge connects two Nodes in a Loop body. Predicate (JMESPath, bool) decides
// whether the edge fires against the envelope emitted by From; omitted
// predicate means always-fire. Transform projects/strips fields from the
// envelope before handing off to To — useful at trust boundaries (e.g. strip
// scope_grants before crossing a2a-remote).
type Edge struct {
	From      string         `json:"from"`
	To        string         `json:"to"`
	Predicate string         `json:"predicate,omitempty"`
	Transform *EdgeTransform `json:"transform,omitempty"`
}

// EdgeTransform is a declarative envelope projection. Day-1 supports:
//
//   - Select: keep only the listed JMESPath-rooted fields
//   - Omit:   drop the listed fields
//   - Merge:  merge a literal JSON object into the envelope
//
// Selecting and omitting are mutually exclusive on a given edge; Merge can
// combine with either.
type EdgeTransform struct {
	Select []string               `json:"select,omitempty"`
	Omit   []string               `json:"omit,omitempty"`
	Merge  map[string]interface{} `json:"merge,omitempty"`
}

// TemplateSnapshot is pinned onto each LoopInstance at creation. The runner
// reads only from the snapshot for the duration of the Loop's life; on-disk
// template changes after creation never affect in-flight instances. Role
// markdown referenced by Nodes does live-bind — see the plan for the
// rationale (templates change freely, roles change deliberately).
type TemplateSnapshot struct {
	Name     string `json:"name"`
	Version  string `json:"version"`
	Hash     string `json:"hash"`
	BodyJSON string `json:"body_json"` // full Loop config at snapshot time
}

// HandoffEnvelope is the unifying handoff primitive. Every Node-to-Node,
// iteration-to-iteration, Loop-to-Loop, and cross-runtime handoff carries one
// of these. The envelope is additive-only — never remove or repurpose a
// field, only add optional ones. SchemaVersion is for cross-runtime
// compatibility checks (e.g. an A2A remote agent declaring "accepts schema
// >=3"); the runner does not do migrations on load.
//
// Embed-with-overflow: blobs over ~100KB go to artifact storage and the
// envelope carries a pointer in ArtifactRefs.
//
// JMESPath-queryable: edge predicates, the convergence predicate, stop
// conditions, join conditions, and the convergence metric all evaluate
// against this shape using the same language.
type HandoffEnvelope struct {
	SchemaVersion int                    `json:"schema_version"`
	LoopID        string                 `json:"loop_id"`
	Iteration     int                    `json:"iteration"`
	FromNode      string                 `json:"from_node"`
	ToNode        string                 `json:"to_node"`
	ArtifactRefs  map[string]interface{} `json:"artifact_refs,omitempty"`
	Observations  map[string]interface{} `json:"observations,omitempty"`
	Findings      map[string]interface{} `json:"findings,omitempty"`
	MemoryRefs    []string               `json:"memory_refs,omitempty"`
	TraceID       string                 `json:"trace_id,omitempty"`
	ScopeGrants   map[string]interface{} `json:"scope_grants,omitempty"`
	ContextCarry  map[string]interface{} `json:"context_carry,omitempty"`
}

// EnvelopeSchemaVersion is the current additive schema version stamped onto
// freshly-emitted envelopes. Bump only when adding fields, never on
// repurpose or removal. Day-1: v1.
const EnvelopeSchemaVersion = 1

// Permission tiers — see plan for semantics. Default applies when a Loop
// declares no tier.
const (
	PermissionInherit = "inherit"
	PermissionDefault = "default"
	PermissionStrict  = "strict"
)

// Node kinds.
const (
	NodeKindAgent    = "agent"
	NodeKindPlaybook = "playbook"
	NodeKindJoin     = "join"
	NodeKindHuman    = "human"
	NodeKindSchedule = "schedule"
)

// Node executors.
const (
	ExecutorHostCLI         = "host-cli"
	ExecutorBrainboxSession = "brainbox-session"
	ExecutorA2ARemote       = "a2a-remote"
)

// Escalation actions.
const (
	EscalationHuman = "human"
	EscalationSkip  = "skip"
	EscalationFail  = "fail"
)

File: brainbox/src/brainbox/loops.py
"""Loop runtime types — HandoffEnvelope, LoopInstance, LoopStatus,
RequiredRef, PermissionTier.

The on-disk loop *definition* lives in ``loop_md`` as ``LoopMarkdown``.
This module holds the runtime shapes the runner reads and writes.

History note: an earlier version held a full ``LoopSpec`` Pydantic
model with embedded JMESPath predicates, Node/Edge graphs, and
StopCondition lists. That format was replaced with the markdown format
(prose stop/escalate sections evaluated by a judge agent each
iteration). See ``loop_md`` for the parser and ``loop_judge`` for the
evaluator.
"""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Constants — kept aligned with app/loops.go
# ---------------------------------------------------------------------------

ENVELOPE_SCHEMA_VERSION = 1


class PermissionTier(str, Enum):
    """How a Loop's worker receives permissions from the profile env.

    Profiles isolate projects; this is the second, finer boundary inside
    one profile that handles the reviewer-vs-worker prompt-injection
    surface within a single loop.
    """

    INHERIT = "inherit"
    DEFAULT = "default"
    STRICT = "strict"


class RequiredRefType(str, Enum):
    """Operator-facing type hint for a loop's required artifact_refs.

    Drives the Trigger form's input rendering (number input for int,
    text input for string, fixed-width font + length hint for sha) and
    documents intent for AI Assist when generating templates. The
    runner does NOT enforce type — the start_loop check is presence-only.
    """

    INT = "int"
    STRING = "string"
    SHA = "sha"


class RequiredRef(BaseModel):
    """Declares an artifact_refs key that the operator (or webhook
    handler) must populate before start_loop accepts the trigger.

    Surfaced in the desktop Trigger tab as a form field; pre-populated
    by the GitHub webhook handler from the PR payload.
    """

    name: str
    type: RequiredRefType = RequiredRefType.STRING
    description: str = ""
    required: bool = True


# ---------------------------------------------------------------------------
# HandoffEnvelope — the unifying handoff primitive
# ---------------------------------------------------------------------------


class HandoffEnvelope(BaseModel):
    """The unifying handoff payload. Every iteration-to-iteration handoff
    and every cross-runtime handoff carries one of these.

    Discipline (do not break):
      - Additive-only schema. Never remove or repurpose a field; only
        add optional ones. ``schema_version`` is stamped for cross-runtime
        compatibility checks (e.g. A2A remote agent declaring
        ``accepts_schema: ">=3"``). The runner does NOT do migrations
        on load.
      - Embed with overflow. Up to ~100 KB embedded; larger blobs go to
        artifact storage and the envelope carries a pointer in
        ``artifact_refs``.
      - Judge-readable. The judge agent reads this shape directly when
        evaluating prose stop/escalate sections, so keep keys descriptive.
    """

    schema_version: int = ENVELOPE_SCHEMA_VERSION
    loop_id: str = ""
    iteration: int = 0
    from_node: str = ""
    to_node: str = ""
    artifact_refs: dict[str, Any] = Field(default_factory=dict)
    observations: dict[str, Any] = Field(default_factory=dict)
    findings: dict[str, Any] = Field(default_factory=dict)
    memory_refs: list[str] = Field(default_factory=list)
    trace_id: str = ""
    scope_grants: dict[str, Any] = Field(default_factory=dict)
    context_carry: dict[str, Any] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# LoopInstance — one loop run
# ---------------------------------------------------------------------------


class LoopStatus(str, Enum):
    """A loop instance's lifecycle. PENDING / RUNNING are active; the rest
    are terminal. CONVERGED is the success exit; the four FAILED-ish
    terminals capture distinct failure shapes for triage.
    """

    PENDING = "pending"
    RUNNING = "running"
    CONVERGED = "converged"
    THRASHING = "thrashing"
    MAX_ITER = "max_iter"
    STOPPED_BY_CONDITION = "stopped_by_condition"  # set when judge says escalate
    FAILED = "failed"
    CANCELLED = "cancelled"


class LoopInstance(BaseModel):
    """One execution of a parsed loop template. Each iteration is
    projected as a child task in the router; this instance holds the
    cross-iteration state the runner needs to decide whether to advance,
    converge, or stop.

    ``template_text`` is the full raw markdown frozen at creation. The
    runner only reads from this snapshot for the loop's life; on-disk
    template changes never affect in-flight instances. ``mermaid`` is
    generated alongside it and survives template edits.

    ``cost_history`` is the per-iteration USD cost series; ``cost_usd``
    is the running total. Used to enforce ``budget_usd`` from the
    template frontmatter.
    """

    id: str
    template_name: str = ""
    template_text: str = ""        # raw markdown — the spec snapshot
    template_hash: str = ""        # short content hash for telemetry
    mermaid: str = ""              # rendered at create time
    parent_task_id: str
    status: LoopStatus = LoopStatus.PENDING
    iteration: int = 0
    envelope: HandoffEnvelope
    cost_history: list[float] = Field(default_factory=list)
    cost_usd: float = 0.0
    current_child_id: str | None = None
    workspace_profile: str | None = None
    created_at: int  # epoch ms
    updated_at: int  # epoch ms
    error: str | None = None
    stop_reason: str | None = None  # judge "reason" string when terminal
